db/db_impl/db_impl_open.cc

```diff
@@ -1929,4 +1929,305 @@ Status DBImpl::Open(const DBOptions& db_options, const std::string& dbname,
   }
   return s;
 }
+
+Status DB::CreateFromDisjointInstances(
+    const MergeInstanceOptions& merge_options, const DBOptions& db_options,
+    const std::string& name,
+    const std::vector<ColumnFamilyDescriptor>& column_families,
+    const std::vector<DB*> instances, std::vector<ColumnFamilyHandle*>* handles,
+    DB** dbptr) {
+  return DBImpl::CreateFromDisjointInstances(merge_options, db_options, name,
+                                             column_families, instances,
+                                             handles, dbptr);
+}
+
+Status DBImpl::CreateFromDisjointInstances(
+    const MergeInstanceOptions& merge_options, const DBOptions& db_options,
+    const std::string& name,
+    const std::vector<ColumnFamilyDescriptor>& column_families,
+    const std::vector<DB*> instances, std::vector<ColumnFamilyHandle*>* handles,
+    DB** dbptr) {
+  Status s;
+  MergeInstanceOptions mopts(merge_options);
+  // Sanity check.
+  if (column_families.size() == 0 ||
+      column_families.front().name !=
+          ROCKSDB_NAMESPACE::kDefaultColumnFamilyName) {
+    return Status::InvalidArgument("Missing default column family descriptor.");
+  }
+  if (!db_options.create_if_missing) {
+    return Status::InvalidArgument(
+        "create_if_missing must be set to create a DB.");
+  }
+  // Create target DB and column families.
+  s = DB::Open(Options(db_options, column_families.front().options), name,
+               dbptr);
+  if (!s.ok()) {
+    delete *dbptr;
+    *dbptr = nullptr;
+    return s;
+  }
+  WriteBufferManager* write_buffer_manager =
+      db_options.write_buffer_manager.get();
+  DBImpl* new_db_impl = static_cast<DBImpl*>(*dbptr);
+  autovector<uint64_t> cf_ids;
+  autovector<ColumnFamilyData*> new_cfds;
+  if (new_db_impl->total_log_size_ > 0) {
+    s = Status::InvalidArgument("Target DB WAL is not empty.");
+  }
+  for (auto& cf : column_families) {
+    ColumnFamilyHandle* cf_handle;
+    if (s.ok()) {
+      if (handles->size() == 0) {
+        auto* cfd = new_db_impl->versions_->GetColumnFamilySet()->GetDefault();
+        cf_handle =
+            new ColumnFamilyHandleImpl(cfd, new_db_impl, &new_db_impl->mutex_);
+      } else {
+        s = new_db_impl->CreateColumnFamily(cf.options, cf.name, &cf_handle);
+      }
+    }
+    if (s.ok()) {
+      assert(cf_handle);
+      handles->push_back(cf_handle);
+      cf_ids.push_back(cf_handle->GetID());
+      ColumnFamilyData* cfd =
+          new_db_impl->versions_->GetColumnFamilySet()->GetColumnFamily(
+              cf_handle->GetID());
+      new_cfds.push_back(cfd);
+      assert(cfd != nullptr && !cfd->IsDropped());
+
+      VersionStorageInfo& vsi = *cfd->current()->storage_info();
+      if (vsi.num_non_empty_levels() != 0) {
+        s = Status::InvalidArgument("Target DB table data is not empty.");
+      }
+    }
+    if (!s.ok()) {
+      break;
+    }
+  }
+  // Validate and prepare source DBs.
+  autovector<DBImpl*> db_impls;
```

Comment:
I'll refactor some of them (e.g. the overlap check can be moved to a pure function). But will keep the rest in place, primarily because they are not pure in the sense that they require a specific context that this function provides, and shouldn't be broken apart to subroutines.

Plus this function is not long in RocksDB's standard. Inline unless it's necessary to refactor is often considered a [good practice](http://number-none.com/blow/blog/programming/2014/09/26/carmack-on-inlined-code.html).